CIP-0031/CIP-0031.md

```diff
@@ -0,0 +1,218 @@
+
+> <sub>NOTE</sub>
+>
+> The following provides an example of the proposed on-chain metadata:
+```json
+{
+  "20": {
+    "<policy-id>": {
+      "<asset-name>": {
+        "ticker": "SUNDAE",
+        "url": "https://dex.sundaeswap.finance/assets/{policy-id}.{asset-name}",
+        "desc": "SUNDAE",
+        "icon": "ipfs://ipfs/<ipfs-hash>",
+        "icon-<size>": "ipfs://ipfs/<ipfs-hash>",
+        "decimals": 6,
+        "version": "1.0"
+      }
+    }
+  }
+}
+```
+
+Field       |          | Description            |
+:---        | :----    | :----                  |
+policy-id   |          | token policy id        |
+asset-name  |          | hex encoded asset name |
+ticker      | OPTIONAL | when present, field and overrides default ticker which is the asset-name |
+url         | OPTIONAL | https only url that refers to metadata stored offchain.  The URL SHOULD use the project domain and MUST return authenticity metadata in either html or json format (see below) |
+desc        | OPTIONAL | additional description that defines the usage of the token |
+icon        | REQUIRED | MUST be either https, ipfs, or data.  icon MUST be a browser supported image format.
+icon-&lt;size&gt; | OPTIONAL | allows teams to provide icon in different sizes. the recommended values of size are 16, 32, 64, 96, 128. icons are assumed to be square so that icon-64 would refer to a 64x64 icon
+decimals    | OPTIONAL | how many decimal places should the token support? For ADA, this would be 6 e.g. 1 ADA is 10^6 Lovelace
+version     | OPTIONAL | when not specified, version will default to `1.0`
+
+> <sub>NOTE</sub>
+>
+> The following provides an example of the alternate proposed on-chain metadata reference:
+
+```json
+{
+  "20": {
+    "<policy-id>": {
+      "<asset-name>": {
+        "ref": "https://assets.sundaeswap.finance",
+      }
+    }
+  }
+}
+```
+
+In some cases, it is not practical or desirable to have embed the metadata on-chain.  In these cases, this proposal allows the on-chain
+
+
+Field       |          | Description            |
+:---        | :----    | :----                  |
+policy-id   |          | token policy id        |
+asset-name  |          | hex encoded asset name |
+ref         | REQUIRED | https only url that holds the metadata in the onchain format (see above).  The URL SHOULD use the project domain and MUST return the token metadata as described above
+
+
+### Authenticity Metadata
+
+To help prevent adversaries from creating fake tokens with the intent to defraud users, projects may authenticate tokens using domains they control.  This
+mechanism provides users a better understanding of where their tokens came from.
+
+The url provided in the token metadata should return content in either html or JSON format.  
+  
+* In JSON format, the url should return a JSON block of the same shape as the on-chain token metadata that includes both the policy-id and asset-name to be verified.
```

Comment:
Maybe a silly question but, if it is required for downstream consumers to query the url provided in the token metadata ... why do we require to include all metadata on-chain in the first place? Wouldn't it be more efficient (space-wise) to only include: (a) a URL, (b) a hash of the expected metadata on-chain, and let client gather other metadata off-chain? 

Also, how does one whether a given URL is valid? Do we expect wallets to whitelist some URLs for known tokens?